actionpack/lib/action_dispatch/request/utils.rb

```diff
@@ -73,6 +77,28 @@ def self.handle_array(params)
           list
         end
       end
+
+      class BinaryParamEncoder < ParamEncoder # :nodoc:
+        def self.encode(request, params)
+          controller = params[:controller] || request.path_parameters[:controller]
```

Comment:
I think we should always use the `path_parameters` since both controller and actions should not come from the GET or POST parameters.